internal/controller/volumegroup/volumegroupreplicationcontent.go

```diff
@@ -0,0 +1,234 @@
+*/
+package volumegroup
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	replicationv1alpha1 "github.com/csi-addons/kubernetes-csi-addons/api/replication.storage/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	ctrl "github.com/ceph/ceph-csi/internal/controller"
+	"github.com/ceph/ceph-csi/internal/rbd"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+)
+
+type ReconcileVGRContent struct {
+	client client.Client
+	config ctrl.Config
+	Locks  *util.VolumeLocks
+}
+
+var (
+	_ reconcile.Reconciler = &ReconcileVGRContent{}
+	_ ctrl.Manager         = &ReconcileVGRContent{}
+)
+
+const (
+	secretNameParameterName      = "replication.storage.openshift.io/group-replication-secret-name"
+	secretNamespaceParameterName = "replication.storage.openshift.io/group-replication-secret-namespace"
+
+	volumeGroupReplicationContentResourceName = "VolumeGroupReplicationContent"
+	volumeGroupReplicationClassResourceName   = "VolumeGroupReplicationClass"
+)
+
+// Init will add the ReconcileVGRContent to the list.
+func Init() {
+	// add ReconcileVGRContent to the list
+	ctrl.ControllerList = append(ctrl.ControllerList, &ReconcileVGRContent{})
+}
+
+// Add adds the newVGRContentReconciler.
+func (r *ReconcileVGRContent) Add(mgr manager.Manager, config ctrl.Config) error {
+	return add(mgr, newVGRContentReconciler(mgr, config))
+}
+
+// newVGRContentReconciler returns a ReconcileVGRContent.
+func newVGRContentReconciler(mgr manager.Manager, config ctrl.Config) reconcile.Reconciler {
+	r := &ReconcileVGRContent{
+		client: mgr.GetClient(),
+		config: config,
+		Locks:  util.NewVolumeLocks(),
+	}
+
+	return r
+}
+
+func ensureCRDsInstalled(mgr manager.Manager) error {
+	ctx := context.Background()
+	vgrcontentlist := &replicationv1alpha1.VolumeGroupReplicationContentList{}
+	if err := mgr.GetAPIReader().List(ctx, vgrcontentlist); err != nil {
+		if strings.Contains(err.Error(), "no matches for kind") {
+			return fmt.Errorf("%s CRD is not installed: %w", volumeGroupReplicationContentResourceName, err)
+		}
+
+		return fmt.Errorf("failed to list %s resources: %w", volumeGroupReplicationContentResourceName, err)
+	}
+
+	vgrclasslist := &replicationv1alpha1.VolumeGroupReplicationClassList{}
+	if err := mgr.GetAPIReader().List(ctx, vgrclasslist); err != nil {
+		if strings.Contains(err.Error(), "no matches for kind") {
+			return fmt.Errorf("%s CRD is not installed: %w", volumeGroupReplicationClassResourceName, err)
+		}
+
+		return fmt.Errorf("failed to list %s resources: %w", volumeGroupReplicationClassResourceName, err)
+	}
```

Comment:
we should not be using list can you please check if we can do something like below

```go
_, err := mgr.GetRESTMapper().RESTMapping(gvk.GroupKind(), gvk.Version)
	if err != nil {
		r.Log.Error(err, "error to get resource", "gvk", gvk)
		return
	}
```

Or check ocs-operator or ocs-client-operator code, we check for CRD existence in both operator as well